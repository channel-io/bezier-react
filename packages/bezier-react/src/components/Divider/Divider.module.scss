@layer components {
  .Divider {
    --b-divider-indent-size: 6px;

    border-radius: var(--radius-1);
    background-color: var(--bdr-black-light);

    &:where(.horizontal) {
      width: calc(100% - 2 * var(--b-divider-indent-size));
      height: var(--b-divider-thickness);
      margin: var(--b-divider-indent-size);

      &:where(.without-indent) {
        width: 100%;
        margin: 0;
      }
      &:where(.without-side-indent) {
        width: 100%;
        margin-left: 0;
        margin-right: 0;
      }
      &:where(.without-parallel-indent) {
        margin-top: 0;
        margin-bottom: 0;
      }
    }

    &:where(.vertical) {
      width: var(--b-divider-thickness);
      height: calc(100% - 2 * var(--b-divider-indent-size));
      margin: var(--b-divider-indent-size);

      &:where(.without-indent) {
        height: 100%;
        margin: 0;
      }
      &:where(.without-side-indent) {
        height: 100%;
        margin-top: 0;
        margin-bottom: 0;
      }
      &:where(.without-parallel-indent) {
        margin-left: 0;
        margin-right: 0;
      }
    }
  }
}
