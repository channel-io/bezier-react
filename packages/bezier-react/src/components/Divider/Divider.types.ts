/* Internal dependencies */
import { BezierComponentProps } from 'Types/ComponentProps'

interface DividerOptions {
  orientation: 'horizontal' | 'vertical'
  withoutSideIndent?: boolean
}

export default interface DividerProps extends
  BezierComponentProps,
  DividerOptions {}

export interface StyledDividerProps extends
  Pick<DividerProps, 'withoutSideIndent' | 'interpolation'> {}
