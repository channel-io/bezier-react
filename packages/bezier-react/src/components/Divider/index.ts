export { default as Divider } from './Divider'
export type { default as DividerProps } from './Divider.types'
