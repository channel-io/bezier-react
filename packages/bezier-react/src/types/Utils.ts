export type Entries<Type> = {
  [Key in keyof Type]: [Key, Type[Key]]
}[keyof Type][]
