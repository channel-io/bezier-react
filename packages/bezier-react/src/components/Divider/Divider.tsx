/* External dependencies */
import React, { forwardRef } from 'react'

/* Internal dependencies */
import DividerProps from './Divider.types'
import * as Styled from './Divider.styled'

export const DIVIDER_TEST_ID = 'bezier-react-divider'

function Divider({
  as,
  testId = DIVIDER_TEST_ID,
  orientation = 'horizontal',
  withoutSideIndent = false,
  ...rest
}: DividerProps,
forwardedRef: React.Ref<HTMLHRElement>,
) {
  const DividerComponent = orientation === 'horizontal'
    ? Styled.HorizontalDivider
    : Styled.VerticalDivider

  return (
    <DividerComponent
      {...rest}
      forwardedAs={as}
      ref={forwardedRef}
      data-testid={testId}
      aria-orientation={orientation}
      withoutSideIndent={withoutSideIndent}
    />
  )
}

export default forwardRef(Divider)
