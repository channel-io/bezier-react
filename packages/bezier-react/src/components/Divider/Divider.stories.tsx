/* External dependencies */
import React from 'react'
import base from 'paths.macro'
import type { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { styled } from 'Foundation'
import { getTitle } from 'Utils/storyUtils'
import { ListItem } from 'Components/ListItem'
import Divider from './Divider'
import DividerProps from './Divider.types'

export default {
  title: getTitle(base),
  component: Divider,
  argTypes: {
    orientation: {
      control: {
        type: 'radio',
        options: ['horizontal', 'vertical'],
      },
    },
  },
} as Meta

interface WrapperProps {
  direction?: 'column' | 'row'
}

const Wrapper = styled.div<WrapperProps>`
  display: flex;
  justify-content: center;
  flex-direction: ${({ direction = 'column' }) => direction};
  width: 200px;
  height: 200px;
`

const Template: Story<DividerProps> = props => (
  <Wrapper>
    <Divider {...props} />
  </Wrapper>
)

export const Primary: Story<DividerProps> = Template.bind({})
Primary.args = {
  orientation: 'horizontal',
  withoutSideIndent: false,
}

const CompositionTemplate: Story<DividerProps> = ({ orientation, ...rest }) => (
  <Wrapper direction={orientation === 'horizontal' ? 'column' : 'row'}>
    <ListItem content="Channel" />
    <Divider
      orientation={orientation}
      {...rest}
    />
    <ListItem content="Bezier" />
  </Wrapper>
)

export const Composition: Story<DividerProps> = CompositionTemplate.bind({})
Composition.args = {
  orientation: 'horizontal',
  withoutSideIndent: false,
}
