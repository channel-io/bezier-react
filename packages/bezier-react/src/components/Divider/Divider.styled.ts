/* Internal dependencies */
import { styled, css } from 'Foundation'
import { StyledDividerProps } from './Divider.types'

const DIVIDER_THICKNESS = 1
const DIVIDER_INDENT_SIZE = 6
const DIVIDER_LENGTH_WITH_SIDE_INDENT = `calc(100% - ${DIVIDER_INDENT_SIZE * 2}px)`

const BaseDivider = styled.hr<StyledDividerProps>`
  box-sizing: content-box;
  margin: ${DIVIDER_INDENT_SIZE}px;
  border: 0;
  border-color: ${({ foundation }) => foundation?.theme?.['bdr-black-light']};
  border-style: solid;

  ${({ interpolation }) => interpolation}
`

export const HorizontalDivider = styled(BaseDivider)`
  width: ${DIVIDER_LENGTH_WITH_SIDE_INDENT};
  border-bottom-width: ${DIVIDER_THICKNESS}px;

  ${({ withoutSideIndent }) => withoutSideIndent && css`
      width: 100%;
      margin-right: 0px;
      margin-left: 0px;
  `}
`

export const VerticalDivider = styled(BaseDivider)`
  height: ${DIVIDER_LENGTH_WITH_SIDE_INDENT};
  border-left-width: ${DIVIDER_THICKNESS}px;

  ${({ withoutSideIndent }) => withoutSideIndent && css`
      height: 100%;
      margin-top: 0px;
      margin-bottom: 0px;
  `}
`
