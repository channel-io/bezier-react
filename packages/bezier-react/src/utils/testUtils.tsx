/* External dependencies */
import React from 'react'
import {
  render as baseRender,
  renderHook as baseRenderHook,
} from '@testing-library/react'

/* Internal dependencies */
import { LightFoundation } from '~/src/foundation'
import BezierProvider from '~/src/providers/BezierProvider'
import { type ChildrenProps } from '~/src/types/ComponentProps'

function TestProviders({ children }: ChildrenProps) {
  return (
    <BezierProvider foundation={LightFoundation}>
      { children }
    </BezierProvider>
  )
}

export function render(
  ui: Parameters<typeof baseRender>[0],
  options?: Parameters<typeof baseRender>[1],
) {
  return baseRender(ui, {
    wrapper: TestProviders,
    ...options,
    legacyRoot: false,
  })
}

interface RenderHookOptions<Props> {
  initialProps?: Props
  wrapper?: React.JSXElementConstructor<{ children: React.ReactElement }>
}

export function renderHook<Result, Props>(
  hook: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props>,
) {
  return baseRenderHook(hook, {
    wrapper: TestProviders,
    ...options,
  })
}
