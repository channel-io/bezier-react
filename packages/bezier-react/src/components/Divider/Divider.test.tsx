/* External dependencies */
import React from 'react'

/* Internal dependencies */
import { render } from 'Utils/testUtils'
import Divider, { DIVIDER_TEST_ID } from './Divider'
import type DividerProps from './Divider.types'

describe('Divider >', () => {
  let props: DividerProps

  beforeEach(() => {
    props = {
      orientation: 'horizontal',
      withoutSideIndent: false,
    }
  })

  const renderDivider = (otherProps?: Partial<DividerProps>) => render(<Divider {...props} {...otherProps} />)

  it('Snapshot >', () => {
    const { getByTestId } = renderDivider()
    const divider = getByTestId(DIVIDER_TEST_ID)

    expect(divider).toMatchSnapshot()
  })

  it('has reset css', () => {
    const { getByTestId } = renderDivider()
    const divider = getByTestId(DIVIDER_TEST_ID)

    expect(divider).toHaveStyle('box-sizing: content-box')
    expect(divider).toHaveStyle('border: 0')
    expect(divider).toHaveStyle('border-style: solid')
  })

  it('has margin', () => {
    const { getByTestId } = renderDivider()
    const divider = getByTestId(DIVIDER_TEST_ID)

    expect(divider).toHaveStyle('margin: 6px')
  })

  it('renders with an aria attribute for web accessibility - horizontal', () => {
    const { getByTestId } = renderDivider()
    const divider = getByTestId(DIVIDER_TEST_ID)

    expect(divider).toHaveAttribute('aria-orientation', 'horizontal')
  })

  it('renders with an aria attribute for web accessibility - vertical', () => {
    const { getByTestId } = renderDivider({ orientation: 'vertical' })
    const divider = getByTestId(DIVIDER_TEST_ID)

    expect(divider).toHaveAttribute('aria-orientation', 'vertical')
  })

  it('renders horizontal divider with correct style', () => {
    const { getByTestId } = renderDivider()
    const divider = getByTestId(DIVIDER_TEST_ID)

    expect(divider).toHaveStyle('width: calc(100% - 12px)')
    expect(divider).toHaveStyle('border-bottom-width: 1px')
  })

  it('renders vertical divider with correct style', () => {
    const { getByTestId } = renderDivider({ orientation: 'vertical' })
    const divider = getByTestId(DIVIDER_TEST_ID)

    expect(divider).toHaveStyle('height: calc(100% - 12px)')
    expect(divider).toHaveStyle('border-left-width: 1px')
  })

  it('renders horizontal divider has not side indent when withoutSideIndent option is true', () => {
    const { getByTestId } = renderDivider({ withoutSideIndent: true })
    const divider = getByTestId(DIVIDER_TEST_ID)

    expect(divider).toHaveStyle('margin-right: 0px')
    expect(divider).toHaveStyle('margin-left: 0px')
  })

  it('renders vertical divider has not side indent when withoutSideIndent option is true', () => {
    const { getByTestId } = renderDivider({ orientation: 'vertical', withoutSideIndent: true })
    const divider = getByTestId(DIVIDER_TEST_ID)

    expect(divider).toHaveStyle('margin-top: 0px')
    expect(divider).toHaveStyle('margin-bottom: 0px')
  })
})
