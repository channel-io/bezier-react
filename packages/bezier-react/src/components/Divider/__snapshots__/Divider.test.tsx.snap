// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider > Snapshot > 1`] = `
.c0 {
  box-sizing: content-box;
  margin: 6px;
  border: 0;
  border-color: #00000014;
  border-style: solid;
}

.c1 {
  width: calc(100% - 12px);
  border-bottom-width: 1px;
}

<hr
  aria-orientation="horizontal"
  class="c0 c1"
  data-testid="bezier-react-divider"
/>
`;
